feat: add C API to `@stdlib/math/base/special/negalucas`

Jordan-Gallivan opened the pull request:
Resolves #893 .

## Description

> What is the purpose of this pull request?

This pull request: Completed the C negative lucas implementation


## Related Issues

> Does this pull request have any related issues?

No.  All tests completed.  All files compiled

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Jordan-Gallivan committed:
C Implementation of negalucas

Jordan-Gallivan committed:
Merge branch 'c_negaluca_implementation' of https://github.com/Jordan-Gallivan/stdlib into c_negaluca_implementation

Jordan-Gallivan commented:
yikes.  Looks like it failed quite a few of the PR checks.  Please let me know what I can do to remedy the above.

kgryte commented:
@Jordan-Gallivan Thanks for this PR. Looking at the CI checks, they don't appear to be due to anything you've done. We're still working out some issues with our CI pipelines. I'll do a more extensive review within the next 24-48 hours.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:
```diff
@@ -1,7 +1,7 @@
 # @license Apache-2.0
 #
-# Copyright (c) 2018 The Stdlib Authors.
+# Copyright (c) 2020 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

Pranavchiku replied:
```suggestion
# Copyright (c) 2018 The Stdlib Authors.
```

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:
```diff
@@ -68,40 +73,54 @@ endif
 	$(QUIET) -rm -f *.o *.out
 
-.PHONY: clean
+.PHONY: clean
```
Are all these changes needed for this PR? I wonder I have not updated any Makefile while adding C implementations.

Jordan-Gallivan replied:
@Pranavchiku I could not get the benchmark to compile with the makefiles that were in the repository.  I pulled the one that was successful from tribonacci and was able to get all files to compile, and all tests to pass

Pranavchiku replied:
Okay, you mixed up two different c benchmarks, create a separate folder `native` inside `c` benchmarks there, and then copy over Makefile.

Pranavchiku replied:
You can take a look at how it works https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/sinh/benchmark/c.

Jordan-Gallivan replied:
that's my bad!  I'll get to fixing it once you all are done with the initial review

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:
```diff
@@ -68,40 +73,54 @@ endif
 	$(QUIET) -rm -f *.o *.out
 
-.PHONY: clean
+.PHONY: clean
```
Add a newline.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+
+    t = tic();
+    for ( i = 0; i < ITERATIONS; i++ ) {
+        x = (int32_t)floor( 77.0*rand_double() );
```
I am not sure whether we should use `floor` here, @kgryte if you can provide your opinion.

Jordan-Gallivan replied:
@Pranavchiku since x is type casted to an int32_t, my assumption is it would act as a floor method, so agreed, the floor is redundant

kgryte replied:
This was addressed.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/example.c:
```diff
@@ -0,0 +1,31 @@
+
+    for ( i = 0; i > -77; i-- ) {
+        v = stdlib_base_negalucas( i );
+        printf( "negalucas(%d) = %lf\n", i, v );
```
```suggestion
        printf( "negalucas(%d) = %d\n", i, v );
```
in my opinion.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/include.gypi:
```diff
@@ -0,0 +1,53 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile:
```diff
@@ -0,0 +1,70 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
You need to update copyright years.
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/addon.c:
```diff
@@ -0,0 +1,22 @@
+#include "stdlib/math/base/special/negalucas.h"
+#include "stdlib/math/base/napi/unary.h"
+
+STDLIB_MATH_BASE_NAPI_MODULE_I_D( stdlib_base_negalucas )
```
Add new line at end.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,118 @@
+* @return      output value
+*
+* @example
+* double out = stdlib_base_negalucas( 1 );
```
Maybe add some more example which returns NaN as well.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,118 @@
+* // returns -1
+*/
+double stdlib_base_negalucas( const int32_t n ) {
+    int32_t an = (int32_t)stdlib_base_abs( n );
```
```suggestion
    int32_t an;
    an = (int32_t)stdlib_base_abs( n );
```

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c:
```diff
@@ -87,10 +87,10 @@ double rand_double() {
 
 /**
-* Computes the nth negaFibonacci number.
+* Computes the nth negaLucas number.
```
Revert this changes.

Pranavchiku reviewed (changes requested):
Dropped a few suggestions kindly apply those and do ping me in case of any difficulties :)

Jordan-Gallivan committed:
Merge branch 'stdlib-js:develop' into c_negaluca_implementation

Jordan-Gallivan committed:
Changes per PR

Jordan-Gallivan commented:
@Pranavchiku made the changes per the above messages.  Just to clarify, the copyright date for all files should be 2023 correct?  If not, I can go through and clean those up.

I took out the floor() method call in native/benchmark.c file per the above discussion.  It looks like it worked.  

I added an example in the src/main.c file that returns NaN.  Didn't know if you wanted an extra new line in between the two examples.  If not, that'll be an easy clean up on my end. 

Additionally, I fixed the MakeFiles for the benchmark and benchmark/native.  That was my bust.

All files compiled and all tests passed!  Please let me know if you see anything else that needs fixing.

Pranavchiku commented:
> Just to clarify, the copyright date for all files should be 2023 correct?

Yes for all those files which you updated and/or added.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:
```diff
@@ -1,7 +1,7 @@
 #/
 # @license Apache-2.0
 #
-# Copyright (c) 2018 The Stdlib Authors.
```
This needs to be reverted, as we are not updating this Makefile.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile:
```diff
@@ -0,0 +1,70 @@
+#/
+clean: clean-addon
+
+.PHONY: clean
```
Add newline at end of file.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,123 @@
+#include "stdlib/math/base/special/negalucas.h"
+#include "stdlib/math/base/special/abs.h"
+
+static const int32_t STDLIB_CONSTANTS_FLOAT64_MAX_SAFE_NTH_NEGALUCAS = 76;
```
Rather than defining it, try using macro at https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/constants/float64/max-safe-nth-lucas.
You may refer https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/src/main.c.

Jordan-Gallivan replied:
@Pranavchiku When i referenced the Max safe nth lucas from the constants, I'm no longer able to get the files to compile.  In the past I've been running the following commands:
make install-node-modules
make examples-c EXAMPLES_FILTER=".*/math/base/special/negalucas/.*"
make benchmark-c BENCHMARKS_FILTER=".*/math/base/special/negalucas/.*"
make install-node-addons NODE_ADDONS_PATTERN=math/base/special/negalucas
make test TESTS_FILTER=".*/math/base/special/negalucas/.*"

is there an intermediate step I need to run after make install-node-modules so the includes can reference constants?

Pranavchiku reviewed (changes requested):
I have left few suggestions kindly have a look at those, along with that `benchmark.native.js` is missing you will need to add that. Thanks for working on this :)

Jordan-Gallivan committed:
Changes per PR

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,124 @@
+#include "stdlib/math/base/special/abs.h"
+#include "stdlib/constants/float64/max_safe_nth_lucas.h"
+
+static const int32_t STDLIB_CONSTANT_FLOAT64_MAX_SAFE_NTH_LUCAS;
```
As you are including "stdlib/constants/float64/max_safe_nth_lucas.h" you do not need to explicitly define `STDLIB_CONSTANT_FLOAT64_MAX_SAFE_NTH_LUCAS`
```suggestion
```

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json:
```diff
@@ -0,0 +1,72 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/special/abs"
```
You will need to add macro to dependency as well.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,124 @@
+    int32_t an;
+    an = (int32_t)stdlib_base_abs( n );
+    if ( n > 0 || an > STDLIB_CONSTANTS_FLOAT64_MAX_SAFE_NTH_NEGALUCAS ) {
+        return 0.0 / 0.0;   // NaN
```
```suggestion
        return 0.0 / 0.0; // NaN
```

Pranavchiku reviewed (changes requested):
After adding MACRO to dependency and removing the declaration, I guess you will be able to clear benchmarks and pass tests. LMK if you face any difficulty.

Jordan-Gallivan committed:
Changes per PR comments on 23 Feb

Jordan-Gallivan commented:
@Pranavchiku just updated the PR with the above changes and all tests passed.  Thanks for your patience, this has been an awesome learning experience.

Pranavchiku committed:
Update lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/Makefile:
```diff
@@ -0,0 +1,146 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

Pranavchiku committed:
Update copyright year.

Pranavchiku commented on lib/node_modules/@stdlib/math/base/special/tribonacci/benchmark/c/Makefile:
```diff
@@ -1,7 +1,7 @@
 # @license Apache-2.0
 #
-# Copyright (c) 2020 The Stdlib Authors.
+# Copyright (c) 2018 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2020 The Stdlib Authors.
```

Pranavchiku committed:
Update lib/node_modules/@stdlib/math/base/special/tribonacci/benchmark/c/Makefile

Pranavchiku reviewed (approved):
LGTM! Thanks @Jordan-Gallivan!! Let @kgryte have a look before merging this.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c:
```diff
@@ -87,10 +87,10 @@ double rand_double() {
 *
-* @param n   negaFibonacci number to compute
-* @return negaFibonacci number
+* @param n  negaLucas number to compute
```
Thanks for catching this and fixing it!

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h:
```diff
@@ -0,0 +1,43 @@
+* limitations under the License.
+*/
+
+/**
+* Header file containing function declarations
+*/
```
```suggestion
```

We can go ahead and remove this, as this originates from files adhering to older project conventions.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,123 @@
+        2918000611027443,
+        -4721424167835364,
+        7639424778862807
+};
```
```suggestion
};

```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,123 @@
+        7639424778862807
+};
+/**
+* Computes the nth Negative Lucas number
```
```suggestion
* Computes the nth negaLucas number.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json:
```diff
@@ -0,0 +1,75 @@
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/unary",
+        "@stdlib/math/base/special/abs",
```
```suggestion
        "@stdlib/math/base/special/labs",
```

We have a 32-bit variant for use in C.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json:
```diff
@@ -0,0 +1,75 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/special/abs",
```
```suggestion
        "@stdlib/math/base/special/labs",
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json:
```diff
@@ -0,0 +1,75 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/special/abs",
```
```suggestion
        "@stdlib/math/base/special/labs",
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,123 @@
+*/
+
+#include "stdlib/math/base/special/negalucas.h"
+#include "stdlib/math/base/special/abs.h"
```
```suggestion
#include "stdlib/math/base/special/labs.h"
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:
```diff
@@ -0,0 +1,123 @@
+* // returns NaN
+*/
+double stdlib_base_negalucas( const int32_t n ) {
+    int32_t an;
+    an = (int32_t)stdlib_base_abs( n );
```
```suggestion
    int32_t an = stdlib_base_labs( n );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+tape( 'if provided negative integers less than `76`, the function returns `NaN`', opts, function test( t ) {
+	var i;
+	var v;
+	for ( i = -77; i < -500; i-- ) {
```
```suggestion
	for ( i = -77; i > -500; i-- ) {
```

As is, this loop will never iterate.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+	t.end();
+});
+
+tape( 'if provided negative integers less than `76`, the function returns `NaN`', opts, function test( t ) {
```
```suggestion
tape( 'if provided nonpositive integers less than `-76`, the function returns `NaN`', opts, function test( t ) {
```

We should maintain parity with `test.js`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+	t.end();
+});
+
+tape( 'the function returns the nth negaLucas number (Lucas relationship)', opts, function test( t ) {
```
```suggestion
tape( 'the function returns the nth negaLucas number', opts, function test( t ) {
```

This test block is not testing the Lucas relationship.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+	var i;
+	for ( i = 0; i > -77; i-- ) {
+		v = negalucas( i );
+		t.strictEqual( v, NEGALUCAS[ i ], 'returns the '+i+'th negaLucas number' );
```
```suggestion
		t.strictEqual( v, NEGALUCAS[ abs( i ) ], 'returns the '+i+'th negaLucas number' );
```

This test is broken, as `i` is nonpositive, but the indices for `NEGALUCAS` must be nonnegative.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+var resolve = require( 'path' ).resolve;
+var tape = require( 'tape' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var tryRequire = require( '@stdlib/utils/try-require' );
```
```suggestion
var pow = require( '@stdlib/math/base/special/pow' );
var abs = require( '@stdlib/math/base/special/abs' );
var tryRequire = require( '@stdlib/utils/try-require' );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,81 @@
+	}
+	t.end();
+});
+
```
```suggestion

tape( 'the function returns the nth negaLucas number (Lucas relationship)', opts, function test( t ) {
	var v;
	var f;
	var i;
	for ( i = 0; i > -77; i-- ) {
		v = negalucas( i );
		f = pow( -1.0, abs(i) ) * lucas( abs(i) );
		t.strictEqual( v, f, 'returns the '+i+'th negaLucas number' );
	}
	t.end();
});

```

Add missing tests for the Lucas relationship. In short, our `test.native.js` files should more or less match our `test.js` files. Meaning, calling into C from JavaScript should yield the same results as if we had just invoked a JavaScript implementation. As such, tests

kgryte committed:
Apply suggestions from code review

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,95 @@
+	var v;
+	var i;
+
+	t.strictEqual( isnan( negalucas( 3.14 ) ), true, 'returns NaN' );
+
```
```suggestion
```

This is one where we don't expect parity between JS and C. C expects 32-bit ints, so we test accordingly.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:
```diff
@@ -0,0 +1,95 @@
+var tape = require( 'tape' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var pow = require( '@stdlib/math/base/special/pow' );
+var abs = require( '@stdlib/math/base/special/abs' );
```
```suggestion
var abs = require( '@stdlib/math/base/special/abs' );
var lucas = require( '@stdlib/math/base/special/lucas' );
```

kgryte committed:
Apply suggestions from code review

kgryte reviewed (approved):
LGTM. Will merge once CI passes. Thanks, @Jordan-Gallivan and @Pranavchiku!

kgryte committed:
Remove comment

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h:
```diff
@@ -0,0 +1,40 @@
+#endif
+
+/**
+* Computes the absolute value of a double-precision floating-point number.
```
```suggestion
* Computes the nth negaLucas number.
```

kgryte committed:
docs: fix description

kgryte reviewed (changes requested):
@Jordan-Gallivan Looking again. Looks like we're missing updates to the README. Would you mind documenting the C APIs in the `negalucas` README file? See Fibonacci for an example.

Jordan-Gallivan commented:
@kgryte I'd be happy to.  Thanks for cleaning up some of the slop in my code!

I noticed you changed from abs to labs.  Would you mind if I ask why (continuing the learning on my end)?

kgryte commented:
Happy to explain.

1. **Avoiding unnecessary type conversions.** The signature for `stdlib_base_abs` is

    ```c
    double stdlib_base_abs( double x );
    ```

    So `double` in and `double` out. The input argument for `stdlib_base_negalucas` is `int32_t`. Accordingly, when passing the input argument to `stdlib_base_abs`, we need to perform an implicit type conversion `(double)<int32_t>`. And then we were performing an explicit type conversion of the return value

    ```c
    int32_t an = (int32_t)stdlib_base_abs(...)
    ```

    In which case, we're performing two type conversions, which is less than optimal.

    In contrast, the signature for `stdlib_base_labs` is

    ```c
    int32_t stdlib_base_labs( int32_t x );
    ```

    which matches our desired input and output dtypes and thus avoids the undesired type conversions.

2. **Avoiding unnecessary branching.** Because of floating-point, `stdlib_base_abs` needs to perform additional branching logic, particularly with regard to positive and negative zero. These concerns are not applicable when working with integer dtypes. Instead, we can perform bit manipulation, which is what `stdlib_base_labs` does and which **can** be easier for a compiler to optimize.

Hope this helps!

Jordan-Gallivan committed:
negalucas readme update

Jordan-Gallivan committed:
Updates to typos

Jordan-Gallivan commented:
@kgryte Thank you so much for the explnation.  I'm really digging what you're doing with this project and it's taught me a a lot so far in the little that I've been able to contribute.  I just updated the README for this and I'll pop back to the tribonacci to get that knocked out as well.

kgryte commented:
@Jordan-Gallivan Glad to help. I've already updated [tribonacci](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/@stdlib/math/base/special/tribonacci#c-apis), so no need to worry about that. Feel free to move another open issue.

kgryte commented on lib/node_modules/@stdlib/math/base/special/negalucas/README.md:
```diff
@@ -148,6 +148,95 @@ for ( i = 0; i > -77; i-- ) {
 
 <!-- /.examples -->
 
+
```
```suggestion
```

kgryte committed:
Remove extra line

kgryte reviewed (approved):
LGTM. Thanks, @Jordan-Gallivan!
